### PR TITLE
HPCC-29115 Add support for uncompressed payloads

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jhtree/jhinplace.cpp [hunk 1]
```diff
@@ -1989,36 +1993,72 @@ bool CInplaceLeafWriteNode::add(offset_t pos, const void * _data, size32_t size,
...
-
+    size32_t required = getDataSize(true);
     bool hasSpace = (required <= maxBytes);
-    if (hasSpace && (keyLen != keyCompareLen))
+    if ((keyLen != keyCompareLen) && ctx.compressionHandler)
     {
-        size32_t oldLzwSize = compressor.buflen();
-        size32_t lzwSpace = maxBytes - (required - oldLzwSize);
-        if (!compressor.adjustLimit(lzwSpace) || !compressor.write(extraData, extraSize))
+        size32_t maxUsedSpace = getDataSize(false) + 1 + (useCompressedPayload || isVariable ? 2 : 0);
```
Comment (richardkchapman) [inline system/jhtree/jhinplace.cpp:2008]: This calculation could use a comment explaining where the values come from...

File: system/jhtree/jhinplace.cpp [hunk 2]
```diff
@@ -1989,36 +1993,72 @@ bool CInplaceLeafWriteNode::add(offset_t pos, const void * _data, size32_t size,
...
         {
-            //Attempting to write data may have flushed some bytes into the lzw output buffer.
-            oldSize += (compressor.buflen() - oldLzwSize);
-            hasSpace = false;
+            if (!useCompressedPayload)
+            {
+                //Reached the threshold where uncompressed data does not fit.
+                //Either need to finish the node or switch to using a compressed payload.
+                unsigned uncompressedSize = uncompressed.length();
+                if (compressor.adjustLimit(maxPayloadSize) && compressor.write(uncompressed.bytes(), uncompressed.length()))
```
Comment (richardkchapman) [inline system/jhtree/jhinplace.cpp:2034]: Could use uncompressedSize in place of uncompressed.length()

**Review by richardkchapman — COMMENTED** [review]
A couple of minor comments.

More generally, could use a description somewhere (in the commit comment? In a comment in the code?) of what the approach that is being taken is, and its limitations. As I understand it, the code here is (a) making sure payload is stored uncompressed if it makes sense to do so, and (b) using the uncompressed buffer as a way to make compression (of the first nodeSize bytes at least) more efficient. But compression of the rest of the payload after this first chunk is still done incrementally.

**Comment by ghalliday** [comment]
@richardkchapman I have added a couple of comments and improved the size calculation - in case node sizes > 16K are ever used.  I will add a more general comment about the approach to the next commit.

**Review by richardkchapman — APPROVED** [review]

**Pull request merged**